TACC/Core-CMS-Custom
# Task/wp 338 339 apcd fix dates except exten

## Overview

Exception [WP-339](https://jira.tacc.utexas.edu/browse/WP-339)

- Address typo
- Make sure expiration date cannot be longer than a year

Extension [WP-338](https://jira.tacc.utexas.edu/browse/WP-338)

1. Requested target date should not be allowed to select beyond a year because extensions cannot be granted for periods longer than one year
2. Current expected due date This should be auto populated from the submitter_calendar based on the data period the user chooses
3. The data period should be a calendar field that the submitter can select from. Note: The day is not needed as adjudicated claims data is identified by month and year.

## Related

[WP-339](https://jira.tacc.utexas.edu/browse/WP-339)
[WP-338](https://jira.tacc.utexas.edu/browse/WP-338)

## Changes

Exception threshold form:

- Make sure the date is restricted so that expiration date field can't be longer than year from today.

Extension form:

- Adjusted styles for the three date fields (requested extension date, applicable data period, and expected due date)
- Adjusted button on extension error page so that it is a button and not link (to pick up styles of the rest of the site)
- Added back requested target date, 
- Added current expected date field that populates with an associated get request. It will get the expected date based on the user's selected applicable data period for the user's submitter id. (This field is uneditable - displays from database as reference and will be pushed to extension DB in the form submission post request)

apcd_database.py:

- Added query to get all applicable data periods that have not already been addressed with APCD admins using the submitter id. The business logic is based on an email provided by Blayton
-  Added query to get current expected due date for that data period based on the submitter id and the applicable data period selected in the form
- Added applicable data period and current expected date back into the create_extension query

## Testing
Extensions

1. Go to [http://localhost:8000/submissions/extension-request/](http://localhost:8000/submissions/extension-request/)
2. Fill out form 
3. When you get to applicable data period, make sure the current expected date populates from a get request once you select something from the auto populated drop down. Add extensions and make sure these fields work after using the add extension button
4. Make sure the requested target date further out from a year can not be selected
5. Check that you can submit the form and check that the post request includes the current_expected_date for each extension you submitted.
6. Go to [http://localhost:8000/administration/list-extensions/](http://localhost:8000/administration/list-extensions/) and make sure your recently submitted extension is in the list

Exceptions

1. Go to [http://localhost:8000/submissions/threshold-exception/](http://localhost:8000/submissions/threshold-exception/)
2. Check that expiration typo is fixed
3. Fill out form 
4. Make sure that you can't select an expiration date further out from a year
5. Add exceptions and make sure you can't select exceptions out further than a year from today in additional exception blocks
6. Check that you can successfully submit form.
7.  Go to [http://localhost:8000/submissions/other-exception/](http://localhost:8000/submissions/other-exception/)
8. Make sure that you can't select an expiration date further out from a year
9. Check that you can successfully submit form.

## UI
Extension: 

![Screenshot 2023-10-31 at 3 41 17 PM](https://github.com/TACC/Core-CMS-Custom/assets/96220951/876e3b84-5924-488c-9908-d7bae55e32ec)

Other Exception - notice how you cannot select dates further from a year from today

![Screenshot 2023-10-31 at 3 51 39 PM](https://github.com/TACC/Core-CMS-Custom/assets/96220951/2509bac5-566b-4467-85fd-affa31b584ae)

Threshold Exception
![Screenshot 2023-10-31 at 3 57 28 PM](https://github.com/TACC/Core-CMS-Custom/assets/96220951/c560e606-5b6c-41be-9bcc-58e1a63e9680)
